From 17d8199ed17e1c557cd7dd43e42c6a34c634ac94 Mon Sep 17 00:00:00 2001
From: Raviteja Sunkara <raviteja@codeaurora.org>
Date: Wed, 14 Sep 2011 13:58:27 +0530
Subject: [PATCH] Make the AccountProjectWatch filters editable.

This allows a user to modify an existing project watch filter.
Every watch is provided with an edit icon which upon clicking, opens an
editor (similar to one where user adds the project watches) which has
projectName and filter fields pre-populated with the existing project
watch and the user can edit and update the fields.

Change-Id: I50aa8a90c1923d4ac2a81ad37c2c1a566ece649b
---
 .../common/data/AccountProjectWatchInfo.java       |    4 +
 .../google/gerrit/common/data/AccountService.java  |    5 +
 .../client/account/AccountProjectWatchDisplay.java |   99 ++++++++++++++++++++
 .../client/account/MyWatchedProjectsScreen.java    |    2 +-
 .../gerrit/client/account/MyWatchesTable.java      |   12 +--
 .../client/account/ProjectWatchKeyEditor.java      |   49 ++++++++--
 .../httpd/rpc/account/AccountServiceImpl.java      |   41 +++++++--
 .../reviewdb/client/AccountProjectWatch.java       |    4 +
 8 files changed, 189 insertions(+), 27 deletions(-)
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/account/AccountProjectWatchDisplay.java

diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountProjectWatchInfo.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountProjectWatchInfo.java
index 581a09b..7b77c62 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountProjectWatchInfo.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountProjectWatchInfo.java
@@ -33,6 +33,10 @@ public final class AccountProjectWatchInfo {
     return watch;
   }
 
+  public void setWatch(AccountProjectWatch w) {
+    watch = w;
+  }
+
   public Project getProject() {
     return project;
   }
diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountService.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountService.java
index 8637223..906f8e1 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountService.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/AccountService.java
@@ -52,6 +52,11 @@ public interface AccountService extends RemoteJsonService {
       AsyncCallback<AccountProjectWatchInfo> callback);
 
   @SignInRequired
+  void updateProjectWatchKey(AccountProjectWatch.Key oldKey,
+      String projectName, String filter,
+      AsyncCallback<AccountProjectWatchInfo> callback);
+
+  @SignInRequired
   void updateProjectWatch(AccountProjectWatch watch,
       AsyncCallback<VoidResult> callback);
 
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/AccountProjectWatchDisplay.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/AccountProjectWatchDisplay.java
new file mode 100644
index 0000000..e5e873f
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/AccountProjectWatchDisplay.java
@@ -0,0 +1,99 @@
+// Copyright (C) 2012 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.gerrit.client.account;
+
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.client.ui.HintTextBox;
+import com.google.gerrit.client.ui.ProjectLink;
+import com.google.gerrit.client.ui.ProjectNameSuggestOracle;
+import com.google.gerrit.common.data.AccountProjectWatchInfo;
+import com.google.gerrit.reviewdb.client.AccountProjectWatch;
+import com.google.gerrit.reviewdb.client.Project;
+import com.google.gerrit.reviewdb.client.Change.Status;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.dom.client.KeyCodes;
+import com.google.gwt.event.dom.client.KeyPressEvent;
+import com.google.gwt.event.dom.client.KeyPressHandler;
+import com.google.gwt.event.logical.shared.SelectionEvent;
+import com.google.gwt.event.logical.shared.SelectionHandler;
+import com.google.gwt.user.client.ui.Button;
+import com.google.gwt.user.client.ui.Composite;
+import com.google.gwt.user.client.ui.FlowPanel;
+import com.google.gwt.user.client.ui.Image;
+import com.google.gwt.user.client.ui.Label;
+import com.google.gwt.user.client.ui.Grid;
+import com.google.gwt.user.client.ui.HTMLTable.CellFormatter;
+import com.google.gwt.user.client.ui.SuggestBox;
+import com.google.gwt.user.client.ui.SuggestBox.DefaultSuggestionDisplay;
+import com.google.gwt.user.client.ui.SuggestOracle.Suggestion;
+
+class AccountProjectWatchDisplay extends Composite {
+  private AccountProjectWatchInfo info;
+  private FlowPanel displayPanel;
+  private ProjectWatchKeyEditor editPanel;
+
+  AccountProjectWatchDisplay(final AccountProjectWatchInfo info) {
+    this.info = info;
+    final FlowPanel fp = new FlowPanel();
+    displayPanel = new FlowPanel();
+
+    final Image editIcon = new Image(Gerrit.RESOURCES.editIcon());
+    editIcon.addClickHandler(new ClickHandler() {
+      @Override
+      public void onClick(final ClickEvent event) {
+        if(editPanel == null) {
+          editPanel = new ProjectWatchKeyEditor(info.getWatch().getKey(), false, true) {
+            public void onCancel() {
+              display();
+              setVisible(false);
+            }
+            public void onUpdate(AccountProjectWatchInfo info) {
+              setWatch(info);
+            }
+          };
+          fp.add(editPanel);
+        }
+        editPanel.setVisible(true);
+      }
+    });
+
+    Grid grid = new Grid(1, 2);
+    grid.setWidget(0, 0, displayPanel);
+    grid.setWidget(0, 1, editIcon);
+
+    fp.add(grid);
+    initWidget(fp);
+    display();
+  }
+
+  private void setWatch(AccountProjectWatchInfo info) {
+    this.info = info;
+    display();
+    editPanel.setVisible(false);
+  }
+
+  private void display() {
+    displayPanel.clear();
+    AccountProjectWatch watch = info.getWatch();
+    displayPanel.add(new ProjectLink(watch.getKey().getProjectName(), Status.NEW));
+    if (watch.getFilter() != null) {
+      Label filter = new Label(watch.getFilter());
+      filter.setStyleName(Gerrit.RESOURCES.css().watchedProjectFilter());
+      displayPanel.add(filter);
+    }
+  }
+}
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java
index 51ccb8a..5bde1e4 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchedProjectsScreen.java
@@ -45,7 +45,7 @@ public class MyWatchedProjectsScreen extends SettingsScreen {
 
     /* top table */
 
-    editor = new ProjectWatchKeyEditor(true, false) {
+    editor = new ProjectWatchKeyEditor(null, true, false) {
       public void onBrowse() {
         displayPopup();
       }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java
index c8465cb..4fba2ab 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/MyWatchesTable.java
@@ -17,10 +17,8 @@ package com.google.gerrit.client.account;
 import com.google.gerrit.client.Gerrit;
 import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.client.ui.FancyFlexTable;
-import com.google.gerrit.client.ui.ProjectLink;
 import com.google.gerrit.common.data.AccountProjectWatchInfo;
 import com.google.gerrit.reviewdb.client.AccountProjectWatch;
-import com.google.gerrit.reviewdb.client.Change.Status;
 import com.google.gwt.event.dom.client.ClickEvent;
 import com.google.gwt.event.dom.client.ClickHandler;
 import com.google.gwt.user.client.DOM;
@@ -123,16 +121,8 @@ public class MyWatchesTable extends FancyFlexTable<AccountProjectWatchInfo> {
   }
 
   protected void populate(final int row, final AccountProjectWatchInfo info) {
-    final FlowPanel fp = new FlowPanel();
-    fp.add(new ProjectLink(info.getProject().getNameKey(), Status.NEW));
-    if (info.getWatch().getFilter() != null) {
-      Label filter = new Label(info.getWatch().getFilter());
-      filter.setStyleName(Gerrit.RESOURCES.css().watchedProjectFilter());
-      fp.add(filter);
-    }
-
     table.setWidget(row, 1, new CheckBox());
-    table.setWidget(row, 2, fp);
+    table.setWidget(row, 2, new AccountProjectWatchDisplay(info));
 
     addNotifyButton(AccountProjectWatch.NotifyType.NEW_CHANGES, info, row, 3);
     addNotifyButton(AccountProjectWatch.NotifyType.ALL_COMMENTS, info, row, 4);
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/ProjectWatchKeyEditor.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/ProjectWatchKeyEditor.java
index 5e27c3a..43516d0 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/ProjectWatchKeyEditor.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/account/ProjectWatchKeyEditor.java
@@ -24,8 +24,8 @@ import com.google.gerrit.client.ui.ProjectNameSuggestOracle;
 import com.google.gerrit.common.data.AccountProjectWatchInfo;
 import com.google.gerrit.common.data.ProjectList;
 import com.google.gerrit.common.PageLinks;
-import com.google.gerrit.reviewdb.AccountProjectWatch;
-import com.google.gerrit.reviewdb.Project;
+import com.google.gerrit.reviewdb.client.AccountProjectWatch;
+import com.google.gerrit.reviewdb.client.Project;
 import com.google.gwt.event.dom.client.ClickEvent;
 import com.google.gwt.event.dom.client.ClickHandler;
 import com.google.gwt.event.dom.client.KeyCodes;
@@ -53,7 +53,7 @@ import com.google.gwtexpui.globalkey.client.GlobalKey;
 import com.google.gwtexpui.globalkey.client.HidePopupPanelCommand;
 import com.google.gwtexpui.user.client.PluginSafeDialogBox;
 
-public class ProjectWatchKeyEditor extends Composite implements
+class ProjectWatchKeyEditor extends Composite implements
     ResizeHandler {
   private HintTextBox projectTxt, filterTxt;
   private Button save, browse, cancel;
@@ -70,7 +70,10 @@ public class ProjectWatchKeyEditor extends Composite implements
   private PopupPanel.PositionCallback popupPosition;
   private HandlerRegistration regWindowResize;
 
-  public ProjectWatchKeyEditor(boolean enableBrowse, boolean enableCancel) {
+  private AccountProjectWatch.Key key;
+
+  ProjectWatchKeyEditor(AccountProjectWatch.Key key,
+      boolean enableBrowse, boolean enableCancel) {
     projectTxt = new HintTextBox();
      final SuggestBox projectBox =
          new SuggestBox(new ProjectNameSuggestOracle(), projectTxt);
@@ -209,6 +212,7 @@ public class ProjectWatchKeyEditor extends Composite implements
     };
 
     initWidget(editPanel);
+    setWatchKey(key);
   }
 
   protected void setPopupPosition(int offsetWidth, int offsetHeight) {
@@ -247,11 +251,10 @@ public class ProjectWatchKeyEditor extends Composite implements
       newFilter = null;
     }
     setEnabled(false);
-    Util.ACCOUNT_SVC.addProjectWatch(newProject, newFilter,
-        new GerritCallback<AccountProjectWatchInfo>() {
+    GerritCallback<AccountProjectWatchInfo> callback = new GerritCallback<AccountProjectWatchInfo>() {
         public void onSuccess(final AccountProjectWatchInfo result) {
+          setWatchKey(result.getWatch().getKey());
           onUpdate(result);
-          setProjectTxt("");
           setEnabled(true);
         }
         @Override
@@ -259,7 +262,16 @@ public class ProjectWatchKeyEditor extends Composite implements
           setEnabled(true);
           super.onFailure(caught);
         }
-    });
+    };
+    if(key != null) {
+      String oldFilter = key.getFilter().get();
+      if (oldFilter.equals(newFilter) && key.getProjectName().equals(newProject)) {
+        return;
+      }
+      Util.ACCOUNT_SVC.updateProjectWatchKey(key, newProject, newFilter, callback);
+    } else {
+      Util.ACCOUNT_SVC.addProjectWatch(newProject, newFilter, callback);
+    }
   }
 
   protected void setEnabled(boolean enable) {
@@ -269,6 +281,11 @@ public class ProjectWatchKeyEditor extends Composite implements
     filterTxt.setEnabled(enable);
   }
 
+  public void setWatchKey(AccountProjectWatch.Key k) {
+    key = k;
+    display();
+  }
+
   public void setSaveTxt(String s) {
     save.setHTML(s);
   }
@@ -285,11 +302,27 @@ public class ProjectWatchKeyEditor extends Composite implements
   }
 
   public void onCancel() {
+    display();
   }
 
   public void onUpdate(AccountProjectWatchInfo info) {
   }
 
+  public void display() {
+    if (key == null) {
+      projectTxt.setText("");
+      filterTxt.setText("");
+    } else {
+      projectTxt.setText(key.getProjectName().get());
+      String filter = key.getFilter().get();
+      if ("*".equals(filter)) {
+        filterTxt.setText(null);
+      } else {
+        filterTxt.setText(filter);
+      }
+    }
+  }
+
   protected void displayPopup() {
     poppingUp = true;
     if (firstPopupLoad) { // For sizing/positioning, delay display until loaded
diff --git a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/account/AccountServiceImpl.java b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/account/AccountServiceImpl.java
index e40df11..eb3e637 100644
--- a/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/account/AccountServiceImpl.java
+++ b/gerrit-httpd/src/main/java/com/google/gerrit/httpd/rpc/account/AccountServiceImpl.java
@@ -152,13 +152,7 @@ class AccountServiceImpl extends BaseServiceImplementation implements
         final ProjectControl ctl = projectControlFactory.validateFor(nameKey);
 
         if (filter != null) {
-          try {
-            ChangeQueryBuilder builder = queryBuilder.create(currentUser.get());
-            builder.setAllowFile(true);
-            builder.parse(filter);
-          } catch (QueryParseException badFilter) {
-            throw new InvalidQueryException(badFilter.getMessage(), filter);
-          }
+          validateFilter(filter);
         }
 
         AccountProjectWatch watch =
@@ -175,6 +169,29 @@ class AccountServiceImpl extends BaseServiceImplementation implements
     });
   }
 
+  public void updateProjectWatchKey(final AccountProjectWatch.Key oldKey,
+      final String projectName, final String filter,
+      final AsyncCallback<AccountProjectWatchInfo> callback) {
+    run(callback, new Action<AccountProjectWatchInfo>() {
+      public AccountProjectWatchInfo run(ReviewDb db) throws OrmException,
+          NoSuchProjectException, InvalidQueryException {
+        final Project.NameKey nameKey = new Project.NameKey(projectName);
+        final ProjectControl ctl = projectControlFactory.validateFor(nameKey);
+
+        if (filter != null) {
+          validateFilter(filter);
+        }
+        AccountProjectWatch.Key newKey = new AccountProjectWatch.Key(
+            ((IdentifiedUser) ctl.getCurrentUser()).getAccountId(),nameKey, filter);
+        AccountProjectWatch watch = db.accountProjectWatches().get(oldKey);
+        watch.setKey(newKey);
+        db.accountProjectWatches().insert(Collections.singleton(watch));
+        db.accountProjectWatches().deleteKeys(Collections.singleton(oldKey));
+        return new AccountProjectWatchInfo(watch, ctl.getProject());
+      }
+    });
+  }
+
   public void updateProjectWatch(final AccountProjectWatch watch,
       final AsyncCallback<VoidResult> callback) {
     if (!getAccountId().equals(watch.getAccountId())) {
@@ -209,4 +226,14 @@ class AccountServiceImpl extends BaseServiceImplementation implements
   public void myAgreements(final AsyncCallback<AgreementInfo> callback) {
     agreementInfoFactory.create().to(callback);
   }
+
+  private void validateFilter(String filter) throws InvalidQueryException {
+    try {
+      ChangeQueryBuilder builder = queryBuilder.create(currentUser.get());
+      builder.setAllowFile(true);
+      builder.parse(filter);
+    } catch (QueryParseException badFilter) {
+      throw new InvalidQueryException(badFilter.getMessage(), filter);
+    }
+  }
 }
diff --git a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/AccountProjectWatch.java b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/AccountProjectWatch.java
index 93e6fb3..726daf5 100644
--- a/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/AccountProjectWatch.java
+++ b/gerrit-reviewdb/src/main/java/com/google/gerrit/reviewdb/client/AccountProjectWatch.java
@@ -120,6 +120,10 @@ public final class AccountProjectWatch {
     return key;
   }
 
+  public void setKey(final AccountProjectWatch.Key k) {
+    key = k;
+  }
+
   public Account.Id getAccountId() {
     return key.accountId;
   }
-- 
1.7.7.1

